ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,335 @@
+        // TODO fill the disk with additional swap segments
+        poolInitialized = true;
+      } catch (Exception e) {
+        throw new StoreException("Exception while initializing DiskSpaceAllocator pool", e,
+            StoreErrorCodes.Initialization_Error);
+      }
+    } finally {
+      reserveFiles.rwLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Allocate a file, that is, take a file matching the requested size from the reserve directory and move it to the
+   * provided file path
+   * @param destinationFile the file path to move the allocated file to.
+   * @param sizeInBytes the size in bytes of the requested file.
+   * @throws IOException
+   */
+  void allocate(File destinationFile, long sizeInBytes) throws IOException {
+    if (!poolInitialized) {
+      LOGGER.debug("Allocating segment before pool is fully initialized to " + destinationFile.getAbsolutePath());
+    }
+    File reserveFile = reserveFiles.remove(sizeInBytes);
+    if (reserveFile == null) {
+      LOGGER.info("Segment of size {} not found in pool; attempting to create a new preallocated file", sizeInBytes);
+      reserveFile = createReserveFile(sizeInBytes);
+    }
+    if (destinationFile.exists() || !reserveFile.renameTo(destinationFile)) {
+      reserveFiles.add(sizeInBytes, reserveFile);
+      throw new IOException("Error while moving reserve file: " + reserveFile.getAbsolutePath() + " to location: "
+          + destinationFile.getAbsolutePath());
+    }
+  }
+
+  /**
+   * Return a file to the pool. The user must keep track of the size of the file allocated.
+   * @param fileToReturn the file to return to the pool.
+   * @param sizeInBytes the size of the file to return.
+   * @throws IOException if the file to return does not exist or cannot be cleaned or recreated correctly.
+   */
+  void free(File fileToReturn, long sizeInBytes) throws IOException {
+    if (!poolInitialized) {
+      LOGGER.debug("Freeing segment before pool is fully initialized");
+    }
+    // For now, we delete the file and create a new one. Newer linux kernel versions support
+    // additional fallocate flags, which will be useful for cleaning up returned files.
+    if (!fileToReturn.delete()) {
+      throw new IOException("Error while cleaning up returned file: " + fileToReturn.getAbsolutePath());
+    }
+    fileToReturn = createReserveFile(sizeInBytes);
+    reserveFiles.add(sizeInBytes, fileToReturn);
+  }
+
+  /**
+   * @return the number of swap segments that should be created in the pool for each segment size.
+   */
+  long getSwapSegmentsPerSize() {
+    return requiredSwapSegmentsPerSize;
+  }
+
+  /**
+   * Inventory existing reserve directories and generate a {@link ReserveFileMap}.
+   * @return a populated {@link ReserveFileMap}
+   */
+  private ReserveFileMap inventoryExistingReserveFiles() {
+    ReserveFileMap reserveFiles = new ReserveFileMap();
+    File[] fileSizeDirs = reserveDir.listFiles(FILE_SIZE_DIR_FILTER);
```

Comment:
is there any clean up required ? If there a chance some file creation failed due to process crashed and some residues are left as a result of it